tools/mason/MasonTest.chpl

```diff
@@ -46,68 +48,84 @@ var files: list(string);
+                               opts=["--parallel"],
+                               defaultValue=false,
+                               flagInversion=false);
+  var updateFlag = parser.addFlag(name="update",
+                                  opts=["--update"],
+                                  flagInversion=true);
+  var setCommOpt = parser.addOption(name="setComm",
+                                 opts=["--setComm"],
+                                 defaultValue="none");
+  // TODO: Why doesn't masonTest support a passthrough for values that should
+  // go to the runtime?
+  var otherArgs = parser.addArgument(name="others", numArgs=0..);
+
+  try! {
+    parser.parseArgs(args);
+  }
+  catch ex : ArgumentError {
+    stderr.writeln(ex.message());
+    masonTestHelp();
+    exit(1);
+  }
+  if helpFlag.valueAsBool() {
+    masonTestHelp();
+    exit(0);
+  }
+
   var skipUpdate = MASON_OFFLINE;
+  var show = showFlag.valueAsBool();
+  var run = !runFlag.valueAsBool();
+  var parallel = parFlag.valueAsBool();
+  keepExec = keepFlag.valueAsBool();
+  subdir = recursFlag.valueAsBool();
+  if updateFlag.hasValue() {
+    skipUpdate = !updateFlag.valueAsBool();
+  }
+  if setCommOpt.hasValue() then setComm = setCommOpt.value();
+
+
   var compopts: list(string);
   var searchSubStrings: list(string);
-  var countArgs = args.indices.low+2;
-  for arg in args[args.indices.low+2..args.indices.high] {
-    countArgs += 1;
-    select (arg) {
-      when '-h'{
-        masonTestHelp();
-        exit(0);
-      }
-      when '--help'{
-        masonTestHelp();
-        exit(0);
-      }
-      when '--show'{
-        show = true;
-      }
-      when '--no-run'{
-        run = false;
-      }
-      when '--parallel' {
-        parallel = true;
-      }
-      when '--' {
-        throw new owned MasonError("Testing does not support -- syntax");
-      }
-      when '--keep-binary' {
-        keepExec = true;
-      }
-      when '--recursive' {
-        subdir = true;
-      }
-      when '--update' {
-        skipUpdate = false;
-      }
-      when '--no-update' {
-        skipUpdate = true;
-      }
-      when '--setComm' {
-        setComm = args[countArgs];
-      }
-      otherwise {
-        if arg.startsWith('--setComm='){
-          setComm = arg['--setComm='.size..];
+
+  if otherArgs.hasValue() {
+    for arg in otherArgs.values() {
+      try! {
+        if isFile(arg) && arg.endsWith(".chpl") {
+          files.append(arg);
         }
-        try! {
-          if isFile(arg) && arg.endsWith(".chpl") {
-            files.append(arg);
-          }
-          else if isDir(arg) {
-            dirs.append(arg);
-          }
-          else if arg.startsWith('-') {
-            compopts.append(arg);
-          }
-          else {
-            searchSubStrings.append(arg);
-          }
+        else if isDir(arg) {
+          dirs.append(arg);
+        }
+        else if arg.startsWith('-') {
+          compopts.append(arg);
```

Comment:
yeah this line is pretty suspect - I think it likely we'll need to change this behavior (to distinguish between giving args to the compile vs the run of the test).

Reply:
>  I think it likely we'll need to change this behavior (to distinguish between giving args to the compile vs the run of the test).

I didn't want to change the command line arguments in this PR, but I agree that this isn't ideal. 

Currently there is no way to specify if the additional options should go to the compiler or the runtime (right now they all appear to go to the compiler), and if you are performing testing outside of a mason package, then no options are used at all because although `runUnitTest` takes a `cmdLineCompopts` argument, it remains unused. A workaround for this case is to set the environment variable `CHPL_MODULE_PATH`, but this is hardly clear or ideal.

Also, running inside a mason package, and trying to specify a module location using the `-M` flag, if the user doesn't use the `=` sign like this `-M=path/to/module` then the logic identifies `path/to/module` as a directory to locate test files in rather than passing `-M path/to/module` to the compiler. I am adding a workaround for this which assumes any value after a flag (as identified by starting with a `-`) is an argument for that flag (unless it itself is a flag), but this still won't solve testing from outside a mason directory because the `compopts` are never used in that case.

Reply:
Might be good to record some of these in future work issues.

Reply:
> I didn't want to change the command line arguments in this PR, but I agree that this isn't ideal.

Yes, to be totally clear, I wasn't trying to ask you to change it in this PR.